crates/environ/src/cranelift.rs

```diff
@@ -318,132 +319,147 @@ fn compile(env: CompileEnv<'_>) -> Result<ModuleCacheDataTupleType, CompileError
+        let mut trap_sink = TrapSink::new();
+        let mut stack_map_sink = StackMapSink::default();
+        context
+            .compile_and_emit(
+                isa,
+                &mut code_buf,
+                &mut reloc_sink,
+                &mut trap_sink,
+                &mut stack_map_sink,
+            )
+            .map_err(|error| {
                 CompileError::Codegen(pretty_error(&context.func, Some(isa), error))
             })?;
 
-            let address_transform = get_function_address_map(&context, input, code_buf.len(), isa);
-
-            let ranges = if env.tunables.debug_info {
-                let ranges = context.build_value_labels_ranges(isa).map_err(|error| {
-                    CompileError::Codegen(pretty_error(&context.func, Some(isa), error))
-                })?;
-                Some(ranges)
-            } else {
-                None
-            };
-
-            Ok((
-                code_buf,
-                context.func.jt_offsets,
-                reloc_sink.func_relocs,
-                address_transform,
-                ranges,
-                context.func.stack_slots,
-                trap_sink.traps,
-                unwind_info,
-                stack_map_sink.finish(),
-            ))
-        })
-        .collect::<Result<Vec<_>, CompileError>>()?
-        .into_iter()
-        .for_each(
-            |(
-                function,
-                func_jt_offsets,
-                relocs,
-                address_transform,
-                ranges,
-                sss,
-                function_traps,
+        let unwind_info = context.create_unwind_info(isa).map_err(|error| {
+            CompileError::Codegen(pretty_error(&context.func, Some(isa), error))
+        })?;
+
+        let address_transform = get_function_address_map(&context, input, code_buf.len(), isa);
+
+        let ranges = if env.tunables.debug_info {
+            let ranges = context.build_value_labels_ranges(isa).map_err(|error| {
+                CompileError::Codegen(pretty_error(&context.func, Some(isa), error))
+            })?;
+            Some(ranges)
+        } else {
+            None
+        };
+
+        Ok((
+            code_buf,
+            context.func.jt_offsets,
+            reloc_sink.func_relocs,
+            address_transform,
+            ranges,
+            context.func.stack_slots,
+            trap_sink.traps,
+            unwind_info,
+            stack_map_sink.finish(),
+        ))
+    };
+
+    let inputs: Vec<FunctionBodyInput> = env.function_body_inputs.into_iter().collect();
+
+    let results: Result<Vec<_>, CompileError> = if env.enable_parallel_compilation {
+        inputs
+            .par_iter()
+            .map_init(FuncTranslator::new, compile_function)
+            .collect()
+    } else {
+        inputs
+            .iter()
+            .map(|input| {
+                let mut func_translator = FuncTranslator::new();
```

Comment:
I think in non-parallel mode this would want to create one `FuncTranslator` on the main thread and use that for all functions?

Reply:
Makes sense, moved it outside the closure.